Does not work but no exception be thrown

## Here is my demo

### HelloService.java

```java
public interface HelloService {

    String sayHello(String name);
}
```

### HelloServiceImpl

```java
public class HelloServiceImpl implements HelloService {

    @Override
    public String sayHello(String name) {
        System.out.println("hello " + name);

        return "hello " + name + " from dubbo";
    }
}
```

### Provider.java

```java
public class Provider {
    public static void main(String[] args) {
        var sc = new ServiceConfig<HelloService>();
        sc.setInterface(HelloService.class);
        sc.setRef(new HelloServiceImpl());

        var rc = new RegistryConfig("zookeeper://127.0.0.1:2181");
        // rc.setTimeout(250000);
        //    rc.setParameters(Map.of(
        //        "blockUntilConnectedWait", "250000"
        //    ));

        var protocolConfig = new ProtocolConfig(CommonConstants.DUBBO, 10880);
        protocolConfig.setSerialization("fury");
        DubboBootstrap.getInstance()
            .application(new ApplicationConfig("provider-demo"))
            .service(sc)
            .registry(rc)
            .protocol(protocolConfig)
            .start()
            .await();
    }
}
```

## Consumer.java

```java
public class Consumer {

    public static void main(String[] args) {
        var referenceConfig = new ReferenceConfig<HelloService>();
        referenceConfig.setInterface(HelloService.class);

        var protocolConfig = new ProtocolConfig(CommonConstants.DUBBO, 10880);
        DubboBootstrap.getInstance()
            .reference(referenceConfig)
            .application(new ApplicationConfig("consumer"))
            .registry(new RegistryConfig("zookeeper://127.0.0.1:2181"))
            .protocol(protocolConfig)
            .start(true);

        System.out.println(referenceConfig.get().sayHello("venus"));
    }
}
```

## Gradle deps

```gradle
dependencies {
    testImplementation platform('org.junit:junit-bom:5.9.1')
    testImplementation 'org.junit.jupiter:junit-jupiter'
    // https://mvnrepository.com/artifact/org.apache.dubbo/dubbo
    implementation 'org.apache.dubbo:dubbo:3.2.5'
    // https://mvnrepository.com/artifact/org.apache.dubbo/dubbo-dependencies-zookeeper
    implementation 'org.apache.dubbo:dubbo-dependencies-zookeeper:3.2.5'
    implementation 'org.furyio:fury-core:0.1.0'
    implementation 'org.furyio:dubbo-serialization-fury:0.2.0-SNAPSHOT'

}
```

**Note: `org.furyio:dubbo-serialization-fury:0.2.0-SNAPSHOT` is my local installed jar**

## Zookeeper env

```shell
docker run -d --name zookeeper -p 2181:2181 --restart=unless-stopped zookeeper:3.8
```

## Result

```log
log4j:WARN No appenders could be found for logger (org.apache.dubbo.rpc.model.FrameworkModel).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Exception in thread "main" org.apache.dubbo.rpc.RpcException: Failed to invoke the method sayHello in the service com.ilxqx.dubbo.demo.HelloService. Tried 3 times of the providers [192.168.20.1:10880] (1/1) from the registry 127.0.0.1:2181 on the consumer 192.168.20.1 using the dubbo version 3.2.5. Last error is: Invoke remote method timeout. method: sayHello, provider: dubbo://192.168.20.1:10880/com.ilxqx.dubbo.demo.HelloService?application=provider-demo&background=false&category=providers,configurators,routers&check=false&deprecated=false&dubbo=2.0.2&dynamic=true&executor-management-mode=isolation&file-cache=true&generic=false&interface=com.ilxqx.dubbo.demo.HelloService&methods=sayHello&pid=92976&prefer.serialization=fury&release=3.2.5&serialization=fury&service-name-mapping=true&side=provider&sticky=false&unloadClusterRelated=false, cause: Timeout after 1000ms waiting for result.
	at org.apache.dubbo.rpc.cluster.support.FailoverClusterInvoker.doInvoke(FailoverClusterInvoker.java:114)
	at org.apache.dubbo.rpc.cluster.support.AbstractClusterInvoker.invoke(AbstractClusterInvoker.java:344)
	at org.apache.dubbo.rpc.cluster.router.RouterSnapshotFilter.invoke(RouterSnapshotFilter.java:46)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.monitor.support.MonitorFilter.invoke(MonitorFilter.java:108)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.rpc.cluster.filter.support.MetricsClusterFilter.invoke(MetricsClusterFilter.java:56)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.rpc.protocol.dubbo.filter.FutureFilter.invoke(FutureFilter.java:52)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.spring.security.filter.ContextHolderParametersSelectedTransferFilter.invoke(ContextHolderParametersSelectedTransferFilter.java:41)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.metrics.filter.MetricsFilter.invoke(MetricsFilter.java:73)
	at org.apache.dubbo.rpc.cluster.filter.support.MetricsConsumerFilter.invoke(MetricsConsumerFilter.java:37)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.rpc.cluster.filter.support.ConsumerClassLoaderFilter.invoke(ConsumerClassLoaderFilter.java:40)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.rpc.cluster.filter.support.ConsumerContextFilter.invoke(ConsumerContextFilter.java:116)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CallbackRegistrationInvoker.invoke(FilterChainBuilder.java:196)
	at org.apache.dubbo.rpc.cluster.support.wrapper.AbstractCluster$ClusterFilterInvoker.invoke(AbstractCluster.java:91)
	at org.apache.dubbo.rpc.cluster.support.wrapper.MockClusterInvoker.invoke(MockClusterInvoker.java:103)
	at org.apache.dubbo.rpc.cluster.support.wrapper.ScopeClusterInvoker.invoke(ScopeClusterInvoker.java:169)
	at org.apache.dubbo.registry.client.migration.MigrationInvoker.invoke(MigrationInvoker.java:284)
	at org.apache.dubbo.rpc.proxy.InvocationUtil.invoke(InvocationUtil.java:61)
	at org.apache.dubbo.rpc.proxy.InvokerInvocationHandler.invoke(InvokerInvocationHandler.java:75)
	at com.ilxqx.dubbo.demo.HelloServiceDubboProxy1.sayHello(HelloServiceDubboProxy1.java)
	at com.ilxqx.dubbo.demo.Consumer.main(Consumer.java:32)
Caused by: java.util.concurrent.TimeoutException: Timeout after 1000ms waiting for result.
	at org.apache.dubbo.rpc.AsyncRpcResult.get(AsyncRpcResult.java:207)
Caused by: java.util.concurrent.TimeoutException: Timeout after 1000ms waiting for result.

	at org.apache.dubbo.rpc.protocol.AbstractInvoker.waitForResultIfSync(AbstractInvoker.java:287)
	at org.apache.dubbo.rpc.protocol.AbstractInvoker.invoke(AbstractInvoker.java:190)
	at org.apache.dubbo.rpc.listener.ListenerInvokerWrapper.invoke(ListenerInvokerWrapper.java:71)
	at org.apache.dubbo.rpc.filter.RpcExceptionFilter.invoke(RpcExceptionFilter.java:41)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334)
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CallbackRegistrationInvoker.invoke(FilterChainBuilder.java:196)
	at org.apache.dubbo.rpc.protocol.ReferenceCountInvokerWrapper.invoke(ReferenceCountInvokerWrapper.java:88)
	at org.apache.dubbo.rpc.cluster.support.AbstractClusterInvoker.invokeWithContext(AbstractClusterInvoker.java:383)
	at org.apache.dubbo.rpc.cluster.support.FailoverClusterInvoker.doInvoke(FailoverClusterInvoker.java:80)
	... 27 more
```
<img width="1517" alt="image" src="https://github.com/fury-project/dubbo-serialization-fury/assets/18739624/47ab02bc-495a-45d6-9ed0-c4b59dfea3f8">

But when I use `fastjson2` or other impl can work:
<img width="1587" alt="image" src="https://github.com/fury-project/dubbo-serialization-fury/assets/18739624/8878e908-916a-40ef-8ad0-3de244e94130">

--- a/src/main/java/org/furyio/serialization/dubbo/BaseFurySerialization.java
+++ b/src/main/java/org/furyio/serialization/dubbo/BaseFurySerialization.java
@@ -1,6 +1,5 @@
 package org.furyio.serialization.dubbo;
 
-import com.google.common.base.Preconditions;
 import io.fury.Fury;
 import io.fury.collection.Tuple2;
 import io.fury.memory.MemoryBuffer;
@@ -27,17 +26,6 @@ public abstract class BaseFurySerialization implements Serialization {
 
   public ObjectInput deserialize(URL url, InputStream input) throws IOException {
     Tuple2<Fury, MemoryBuffer> tuple2 = getFury();
-    Fury fury = tuple2.f0;
-    MemoryBuffer buffer = tuple2.f1;
-    buffer.readerIndex(0);
-    int readOffset = input.read(buffer.getHeapMemory(), 0, 4);
-    Preconditions.checkArgument(readOffset == 4, readOffset);
-    int size = buffer.readInt();
-    int end = 4 + size;
-    buffer.ensure(end);
-    while (readOffset < end) {
-      readOffset += input.read(buffer.getHeapMemory(), readOffset, end - readOffset);
-    }
-    return new FuryObjectInput(fury, buffer);
+    return new FuryObjectInput(tuple2.f0, tuple2.f1, input);
   }
 }

--- a/src/main/java/org/furyio/serialization/dubbo/FuryObjectInput.java
+++ b/src/main/java/org/furyio/serialization/dubbo/FuryObjectInput.java
@@ -2,6 +2,8 @@ package org.furyio.serialization.dubbo;
 
 import io.fury.Fury;
 import io.fury.memory.MemoryBuffer;
+import java.io.IOException;
+import java.io.InputStream;
 import java.lang.reflect.Type;
 import org.apache.dubbo.common.serialize.ObjectInput;
 
@@ -9,15 +11,17 @@ import org.apache.dubbo.common.serialize.ObjectInput;
 public class FuryObjectInput implements ObjectInput {
   private final Fury fury;
   private final MemoryBuffer buffer;
+  private final InputStream input;
 
-  public FuryObjectInput(Fury fury, MemoryBuffer buffer) {
+  public FuryObjectInput(Fury fury, MemoryBuffer buffer, InputStream input) {
     this.fury = fury;
     this.buffer = buffer;
+    this.input = input;
   }
 
   @Override
   public Object readObject() {
-    return fury.deserializeJavaObjectAndClass(buffer);
+    return fury.deserializeJavaObjectAndClass(input);
   }
 
   @Override
@@ -31,42 +35,53 @@ public class FuryObjectInput implements ObjectInput {
   }
 
   @Override
-  public boolean readBool() {
-    return buffer.readBoolean();
+  public boolean readBool() throws IOException {
+    readBytes(buffer.getHeapMemory(), 1);
+    return buffer.getBoolean(0);
   }
 
   @Override
-  public byte readByte() {
-    return buffer.readByte();
+  public byte readByte() throws IOException {
+    readBytes(buffer.getHeapMemory(), 1);
+    return buffer.get(0);
   }
 
   @Override
-  public short readShort() {
-    return buffer.readShort();
+  public short readShort() throws IOException {
+    readBytes(buffer.getHeapMemory(), 2);
+    return buffer.getShort(0);
   }
 
   @Override
-  public int readInt() {
-    return buffer.readVarInt();
+  public int readInt() throws IOException {
+    readBytes(buffer.getHeapMemory(), 4);
+    return buffer.getInt(0);
   }
 
   @Override
-  public long readLong() {
-    return buffer.readLong();
+  public long readLong() throws IOException {
+    readBytes(buffer.getHeapMemory(), 8);
+    return buffer.getLong(0);
   }
 
   @Override
-  public float readFloat() {
-    return buffer.readFloat();
+  public float readFloat() throws IOException {
+    readBytes(buffer.getHeapMemory(), 4);
+    return buffer.getFloat(0);
   }
 
   @Override
-  public double readDouble() {
-    return buffer.readDouble();
+  public double readDouble() throws IOException {
+    readBytes(buffer.getHeapMemory(), 8);
+    return buffer.getDouble(0);
   }
 
   @Override
-  public String readUTF() {
+  public String readUTF() throws IOException {
+    int size = readInt();
+    buffer.readerIndex(0);
+    buffer.ensure(size);
+    readBytes(buffer.getHeapMemory(), size);
     if (buffer.readBoolean()) {
       return fury.readJavaString(buffer);
     } else {
@@ -75,11 +90,17 @@ public class FuryObjectInput implements ObjectInput {
   }
 
   @Override
-  public byte[] readBytes() {
-    if (buffer.readBoolean()) {
-      return buffer.readBytesWithSizeEmbedded();
-    } else {
-      return null;
+  public byte[] readBytes() throws IOException {
+    int size = readInt();
+    byte[] bytes = new byte[size];
+    readBytes(bytes, size);
+    return bytes;
+  }
+
+  private void readBytes(byte[] bytes, int size) throws IOException {
+    int off = 0;
+    while (off != size) {
+      off += input.read(bytes, off, size - off);
     }
   }
 }

--- a/src/main/java/org/furyio/serialization/dubbo/FuryObjectOutput.java
+++ b/src/main/java/org/furyio/serialization/dubbo/FuryObjectOutput.java
@@ -1,6 +1,5 @@
 package org.furyio.serialization.dubbo;
 
-import com.google.common.base.Preconditions;
 import io.fury.Fury;
 import io.fury.memory.MemoryBuffer;
 import java.io.IOException;
@@ -21,79 +20,72 @@ public class FuryObjectOutput implements ObjectOutput {
     this.fury = fury;
     this.buffer = buffer;
     this.output = output;
-    if (buffer.writerIndex() != 0) {
-      throw new IllegalArgumentException("Index should be 0 instead of " + buffer.writerIndex());
-    }
-    buffer.writeInt(-1);
   }
 
   public void writeObject(Object obj) {
-    fury.serializeJavaObjectAndClass(buffer, obj);
+    fury.serializeJavaObjectAndClass(output, obj);
   }
 
-  public void writeBool(boolean v) {
-    buffer.writeBoolean(v);
+  public void writeBool(boolean v) throws IOException {
+    buffer.unsafePutBoolean(0, v);
+    output.write(buffer.getHeapMemory(), 0, 1);
   }
 
-  public void writeByte(byte v) {
-    buffer.writeByte(v);
+  public void writeByte(byte v) throws IOException {
+    buffer.unsafePut(0, v);
+    output.write(buffer.getHeapMemory(), 0, 1);
   }
 
-  public void writeShort(short v) {
-    buffer.writeShort(v);
+  public void writeShort(short v) throws IOException {
+    buffer.unsafePutShort(0, v);
+    output.write(buffer.getHeapMemory(), 0, 2);
   }
 
-  public void writeInt(int v) {
-    buffer.writeVarInt(v);
+  public void writeInt(int v) throws IOException {
+    buffer.unsafePutInt(0, v);
+    output.write(buffer.getHeapMemory(), 0, 4);
   }
 
-  public void writeLong(long v) {
-    buffer.writeLong(v);
+  public void writeLong(long v) throws IOException {
+    buffer.unsafePutLong(0, v);
+    output.write(buffer.getHeapMemory(), 0, 8);
   }
 
-  public void writeFloat(float v) {
-    buffer.writeFloat(v);
+  public void writeFloat(float v) throws IOException {
+    buffer.unsafePutFloat(0, v);
+    output.write(buffer.getHeapMemory(), 0, 4);
   }
 
-  public void writeDouble(double v) {
-    buffer.writeDouble(v);
+  public void writeDouble(double v) throws IOException {
+    buffer.unsafePutDouble(0, v);
+    output.write(buffer.getHeapMemory(), 0, 8);
   }
 
-  public void writeUTF(String v) {
+  public void writeUTF(String v) throws IOException {
+    // avoid `writeInt` overwrite sting data.
+    buffer.writerIndex(4);
     if (v != null) {
       buffer.writeBoolean(true);
       fury.writeJavaString(buffer, v);
     } else {
       buffer.writeBoolean(false);
     }
+    int size = buffer.writerIndex() - 4;
+    writeInt(size);
+    output.write(buffer.getHeapMemory(), 4, size);
   }
 
-  public void writeBytes(byte[] v) {
-    if (v != null) {
-      buffer.writeBoolean(true);
-      buffer.writeBytesWithSizeEmbedded(v);
-    } else {
-      buffer.writeBoolean(false);
-    }
+  public void writeBytes(byte[] v) throws IOException {
+    writeInt(v.length);
+    output.write(v);
   }
 
-  public void writeBytes(byte[] v, int off, int len) {
-    if (v != null) {
-      buffer.writeBoolean(true);
-      buffer.writePositiveVarInt(len);
-      buffer.writeBytes(v, off, len);
-    } else {
-      buffer.writeBoolean(false);
-    }
+  public void writeBytes(byte[] v, int off, int len) throws IOException {
+    writeInt(len);
+    output.write(v, off, len);
   }
 
   public void flushBuffer() throws IOException {
-    byte[] heapMemory = buffer.getHeapMemory();
-    Preconditions.checkNotNull(heapMemory);
-    final int targetIndex = buffer.unsafeHeapReaderIndex();
-    buffer.putInt(0, buffer.writerIndex() - 4);
-    output.write(heapMemory, targetIndex, buffer.writerIndex());
-    buffer.writerIndex(0);
     output.flush();
   }
 }

